frame/staking/src/pallet/impls.rs

```diff
@@ -657,18 +666,17 @@ impl<T: Config> Pallet<T> {
 	/// - after a `withdraw_unbonded()` call that frees all of a stash's bonded balance.
 	/// - through `reap_stash()` if the balance has fallen to zero (through slashing).
 	pub(crate) fn kill_stash(stash: &T::AccountId, num_slashing_spans: u32) -> DispatchResult {
-		let controller = <Bonded<T>>::get(stash).ok_or(Error::<T>::NotStash)?;
-
-		slashing::clear_stash_metadata::<T>(stash, num_slashing_spans)?;
+		slashing::clear_stash_metadata::<T>(&stash, num_slashing_spans)?;
 
-		<Bonded<T>>::remove(stash);
-		<Ledger<T>>::remove(&controller);
+		// removes controller from `Bonded` and staking ledger from `Ledger`.
+		StakingLedger::<T>::kill(&stash)?;
+		<Bonded<T>>::remove(&stash);
+		<Payee<T>>::remove(&stash);
```

Comment:
I would love this function to not contain 

```
		<Bonded<T>>::remove(&stash);
		<Payee<T>>::remove(&stash);
```

as the outcome of this PR.

Reply:
Yes, I think that makes sense. 

The `<Bonded<T>>::remove(&stash);` should not have lingered outside the staking ledger API and it trivial to also mutate/read the reward preferences in the staking ledger API. on it.